File: packages/cli/src/commands/import.ts
/**
 * import - Run import pipeline.
 */

import type { ImportResult } from '@fin/core';
import { importInbox } from '@fin/core';
import { initConfig } from '@fin/core/config';
import { defineCommand } from 'citty';

import { resolveDbPath } from '../db';
import { json, log } from '../logger';

function renderProcessedFiles(files: string[]): void {
	if (files.length === 0) return;
	log('Processed:');
	for (const file of files) {
		log(`  ${file}`);
	}
	log('');
}

function renderSkippedFiles(files: Array<{ path: string; reason: string }>): void {
	if (files.length === 0) return;
	log('Skipped:');
	for (const { path, reason } of files) {
		log(`  ${path}: ${reason}`);
	}
	log('');
}

function renderUnmappedDescriptions(descriptions: string[]): void {
	if (descriptions.length === 0) return;
	log(`\nUnmapped descriptions: ${descriptions.length}`);
	for (const desc of descriptions.slice(0, 10)) {
		log(`  "${desc}"`);
	}
	if (descriptions.length > 10) {
		log(`  ... and ${descriptions.length - 10} more`);
	}
}

function renderTableOutput(result: ImportResult): void {
	renderProcessedFiles(result.processedFiles);
	renderSkippedFiles(result.skippedFiles);

	log('Results:');
	log(`  Transactions parsed: ${result.totalTransactions}`);
	if (result.duplicateTransactions > 0) {
		log(`  Duplicates skipped: ${result.duplicateTransactions}`);
	}
	log(`  Journal entries attempted: ${result.journalEntriesAttempted}`);
	log(`  Journal entries created: ${result.journalEntriesCreated}`);
	if (result.transferPairsCreated > 0) {
		log(`  Transfer pairs created: ${result.transferPairsCreated}`);
	}
	if (result.entryErrors.length > 0) {
		log(`  Entry errors: ${result.entryErrors.length}`);
		for (const err of result.entryErrors.slice(0, 10)) {
			log(`    ${err}`);
		}
		if (result.entryErrors.length > 10) {
			log(`    ... and ${result.entryErrors.length - 10} more`);
		}
	}

	if (result.accountsTouched.length > 0) {
		log(`  Accounts touched: ${result.accountsTouched.join(', ')}`);
	}

	renderUnmappedDescriptions(result.unmappedDescriptions);

	if (result.archivedFiles.length > 0) {
		log(`\nArchived ${result.archivedFiles.length} file${result.archivedFiles.length !== 1 ? 's' : ''}`);
		for (const file of result.archivedFiles.slice(0, 5)) {
			log(`  ${file}`);
		}
		if (result.archivedFiles.length > 5) {
			log(`  ... and ${result.archivedFiles.length - 5} more`);
		}
	}
}

export const importCmd = defineCommand({
	meta: { name: 'import', description: 'Import transactions from inbox' },
	args: {
		inbox: { type: 'string', description: 'Custom inbox directory' },
		format: { type: 'string', description: 'Output format: table, json', default: 'table' },
		db: { type: 'string', description: 'Database path' },
	},
	async run({ args }) {
		initConfig();

		const format = args.format === 'json' ? 'json' : 'table';
		const inboxDir = args.inbox;
		const dbPath = resolveDbPath(args.db ? { options: new Map([['db', args.db]]) } : undefined);

		if (format !== 'json') {
			log('Scanning inbox...\n');
		}

		const options: { inboxDir?: string; dbPath: string; migrate: boolean } = { dbPath, migrate: true };
		if (inboxDir) options.inboxDir = inboxDir;
		const result = await importInbox(options);

		if (format === 'json') {
			json(result);
			return;
		}

		renderTableOutput(result);
	},
});

File: crates/fin-cli/src/commands/import.rs
use serde_json::json;

use fin_sdk::{ImportInboxOptions, ImportMode, import_inbox};

use crate::commands::{CommandFailure, CommandResult, map_fin_error};
use crate::envelope::MetaExtras;
use crate::error::ExitCode;

pub fn run(
    inbox: Option<&str>,
    db: Option<&str>,
    full_export: bool,
) -> Result<CommandResult, CommandFailure> {
    let mode = if full_export {
        ImportMode::FullExport
    } else {
        ImportMode::Append
    };
    let result = import_inbox(ImportInboxOptions {
        inbox_dir: inbox.map(std::path::PathBuf::from),
        db_path: db.map(std::path::PathBuf::from),
        migrate: true,
        mode,
        ..ImportInboxOptions::default()
    })
    .map_err(|error| map_fin_error("import", error))?;

    let text = format!(
        "Results:\n  Mode: {:?}\n  Transactions parsed: {}\n  Duplicates skipped: {}\n  Replaced provider transactions: {}\n  Journal entries created: {}\n  Transfer pairs created: {}\n  Entry errors: {}\n  Archived files: {}",
        result.mode,
        result.total_transactions,
        result.duplicate_transactions,
        result.replaced_provider_transactions,
        result.journal_entries_created,
        result.transfer_pairs_created,
        result.entry_errors.len(),
        result.archived_files.len(),
    );

    Ok(CommandResult {
        tool: "import",
        data: json!({
            "mode": result.mode,
            "processedFiles": result.processed_files,
            "archivedFiles": result.archived_files,
            "skippedFiles": result.skipped_files,
            "totalTransactions": result.total_transactions,
            "uniqueTransactions": result.unique_transactions,
            "duplicateTransactions": result.duplicate_transactions,
            "journalEntriesAttempted": result.journal_entries_attempted,
            "journalEntriesCreated": result.journal_entries_created,
            "transferPairsCreated": result.transfer_pairs_created,
            "replacedProviderTransactions": result.replaced_provider_transactions,
            "replacedJournalEntries": result.replaced_journal_entries,
            "entryErrors": result.entry_errors,
            "accountsTouched": result.accounts_touched,
            "unmappedDescriptions": result.unmapped_descriptions,
        }),
        text,
        meta: MetaExtras {
            count: Some(result.journal_entries_created),
            total: None,
            has_more: None,
        },
        exit_code: ExitCode::Success,
    })
}
